In [24]:
import re
import math


# Loading data

In [19]:
file_path = 'data/input.txt'  # Replace with the actual file path
with open(file_path, 'r') as file:
    file_data = file.readlines()

file_data = [line.rstrip() for line in file_data]

print(f'Number of rows : {len(file_data)}, data type: {type(file_data)} number of characters {len(file_data[0])}')
file_data


Number of rows : 140, data type: <class 'list'> number of characters 140


['...788.............................54.........501...........555.........270.................................521......893....................',
 '..../..*963........................*..860......................*....53...../.....................52.................&....347........428*522.',
 '............*......41..481+.......462....$..187......678.......420....-....................&115.+...........................+...............',
 '............707....&.........562...........*..........*.....................438....................&877..660....199..145.........71.........',
 '.....210................356..*.........977.68.........38.......835.622.332.....*300.....131.422..............89..*.....+..........$.........',
 '..............14..312......+..926.....*.......529..*............*...*....*.............*......%...310.......*...835..................885....',
 '...416../467..........................423.....*...143...132..955...356...124.........588..947....*.....512......................

# Part 1

--- Day 3: Gear Ratios ---

You and the Elf eventually reach a gondola lift station; he says the gondola lift will take you up to the water source, but this is as far as he can bring you. You go inside.

It doesn't take long to find the gondolas, but there seems to be a problem: they're not moving.

"Aaah!"

You turn around to see a slightly-greasy Elf with a wrench and a look of surprise. "Sorry, I wasn't expecting anyone! The gondola lift isn't working right now; it'll still be a while before I can fix it." You offer to help.

The engineer explains that an engine part seems to be missing from the engine, but nobody can figure out which one. If you can add up all the part numbers in the engine schematic, it should be easy to work out which part is missing.

The engine schematic (your puzzle input) consists of a visual representation of the engine. There are lots of numbers and symbols you don't really understand, but apparently any number adjacent to a symbol, even diagonally, is a "part number" and should be included in your sum. (Periods (.) do not count as a symbol.)

Here is an example engine schematic:


467..114..

...*......

..35..633.

......#...

617*......

.....+.58.

..592.....

......755.

...$.*....

.664.598..

In this schematic, two numbers are not part numbers because they are not adjacent to a symbol: 114 (top right) and 58 (middle right). Every other number is adjacent to a symbol and so is a part number; their sum is 4361.

Of course, the actual engine schematic is much larger. What is the sum of all of the part numbers in the engine schematic?

Your puzzle answer was 554003.

The first half of this puzzle is complete! It provides one gold star: *

In [22]:
def parse_input(data):
    nums = {} # Dictionary to store numbers with their line number and index
    syms = {} # Dictionary to store symbols with their line number and index
    idx_num = 0 # Unique identifier for each number

    for r, line in enumerate(data):
        # Replace non-digit characters with spaces and split to find all numbers
        line_nums = re.sub(r"\D", " ", line).split()
        offset = 0
        for n in line_nums:
            pos = line.index(n, offset) # Find the start index of each number
            for step in range(len(n)): # For each digit in the number
                nums[(r, pos + step)] = (int(n), idx_num) # Store the entire number and its index
            offset = pos + len(n) # Update offset for next search
            idx_num += 1

        # Replace digits and periods with spaces and split to find all symbols
        line_syms = re.sub(r"[\d\.]", " ", line).split()
        offset = 0
        for sym in line_syms:
            pos = line.index(sym, offset)  # Find the index of each symbol
            syms[(r, pos)] = sym # Store the symbol
            offset = pos + 1

    return nums, syms


554003

In [ ]:
def part1(data):
    # Parse the input data to separate numbers and symbols
    nums, symbols = parse_input(data)
    adj_nums = [] # List to store adjacent numbers to symbols

    for pos, _ in symbols.items():
        r, c = pos
        # Generate positions adjacent (including diagonals) to the symbol
        adj_pos = [(r + x, c + y) for x in [-1, 0, 1] for y in [-1, 0, 1]]
        # Extend the list with numbers found at these adjacent positions
        adj_nums.extend([nums[pos] for pos in adj_pos if pos in nums])

    # Return the sum of unique numbers found adjacent to symbols
    return sum(item[0] for item in set(adj_nums))


In [ ]:
# Now let's use the part1 function to solve the puzzle with the given input data
result_part1 = part1(file_data)
result_part1


# Part two

--- Part Two ---

The engineer finds the missing part and installs it in the engine! As the engine springs to life, you jump in the closest gondola, finally ready to ascend to the water source.

You don't seem to be going very fast, though. Maybe something is still wrong? Fortunately, the gondola has a phone labeled "help", so you pick it up and the engineer answers.

Before you can explain the situation, she suggests that you look out the window. There stands the engineer, holding a phone in one hand and waving with the other. You're going so slowly that you haven't even left the station. You exit the gondola.

The missing part wasn't the only issue - one of the gears in the engine is wrong. A gear is any * symbol that is adjacent to exactly two part numbers. Its gear ratio is the result of multiplying those two numbers together.

This time, you need to find the gear ratio of every gear and add them all up so that the engineer can figure out which gear needs to be replaced.

Consider the same engine schematic again:

467..114..

...*......

..35..633.

......#...

617*......

.....+.58.

..592.....

......755.

...$.*....

.664.598..

In this schematic, there are two gears. The first is in the top left; it has part numbers 467 and 35, so its gear ratio is 16345. The second gear is in the lower right; its gear ratio is 451490. (The * adjacent to 617 is not a gear because it is only adjacent to one part number.) Adding up all of the gear ratios produces 467835.

What is the sum of all of the gear ratios in your engine schematic?

In [28]:
def part2(data):
    nums, symbols = parse_input(data)
    total_gear_ratios = 0

    for pos, symbol in symbols.items():
        if symbol == "*":
            r, c = pos
            adj_pos = [(r + x, c + y) for x in [-1, 0, 1] for y in [-1, 0, 1]]
            adj_nums = set([nums[pos] for pos in adj_pos if pos in nums])
            if len(adj_nums) == 2:
                total_gear_ratios += math.prod([item[0] for item in adj_nums])

    return total_gear_ratios


In [29]:
result_part2 = part2(file_data)
result_part2


87263515